File: src/component/MapComponent.java
package component;

import org.newdawn.slick.Color;
//import org.newdawn.slick.Font;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.gui.GUIContext;


import model.WorldMap;
import model.worldMap.*;


/**
 * This component will hold the cities and the trails for the map 
 *
 */
public class MapComponent extends Component {
	private WorldMap worldMap;
	private Boolean devMode;
	
	public MapComponent(GUIContext context, int width, int height){
		super(context, width, height);
		
	}
	
	public MapComponent(GUIContext context, WorldMap worldMap, Boolean devMode){
		super(context, context.getWidth(), context.getHeight());
		this.worldMap = worldMap;
		this.devMode = devMode;
		
	}
	
	/**
	 * this will draw a trail from a startX,startY to endX,endY
	 * 
	 * @param startX
	 * @param startY
	 * @param endX
	 * @param endY
	 * @param node the Node this trail originated in
	 * @param trail the trail we are painting
	 */
	private void drawIndividualTrail(Graphics g, double startX, double startY, double endX, double endY, LocationNode node, TrailEdge trail){
		g.setLineWidth(3);
		Color startColor;
		Color endColor;
		double colorMult = (node.getRank()/(1.0 * worldMap.getMaxRank()));
		if (trail.isTaken()) {
			startColor = new Color(0, 255, 255);
			endColor = new Color(0,255,255);				
		} else {
			startColor = new Color((int) (100 * colorMult), (int)(50 * colorMult), (int)(30 * colorMult));
			endColor = new Color((int) (255 * colorMult), (int)(128 * colorMult), (int)(64 * colorMult));			
		}
		g.drawGradientLine((float)startX + 5, (float)startY + 5, startColor, (float) endX + 5, (float) endY + 5, endColor);
	}


	@Override
	public void render(GUIContext context, Graphics g) throws SlickException {
			for (int rank = 0; rank <= worldMap.MAX_RANK; rank++){
				for(LocationNode node : worldMap.getMapNodes().get(rank)) {
					for (int nodeTrail = 0; nodeTrail < node.getTrails(); nodeTrail++){
						if ((node.getOutBoundTrailByIndex(nodeTrail).isVisible()) || this.devMode)  {
						drawIndividualTrail(g, node.getOutBoundTrailByIndex(nodeTrail).getOrigin().getPlayerMapX(),
							node.getOutBoundTrailByIndex(nodeTrail).getOrigin().getPlayerMapY(),
							node.getOutBoundTrailByIndex(nodeTrail).getDestination().getPlayerMapX(),
							node.getOutBoundTrailByIndex(nodeTrail).getDestination().getPlayerMapY(), node, node.getOutBoundTrailByIndex(nodeTrail));
						}//if visible draw it
					}//iTrails for
				}//for location node in node lists
			}//for i <= maxrank
			
			super.render(context,g);
	}
	


}

File: web/src/component/MapComponent.ts
import { Component } from './Component';
import { WorldMap } from '../model/WorldMap';
import { LocationNode } from '../model/LocationNode';
import { TrailEdge } from '../model/TrailEdge';
import { Color } from '../core/Color';

/**
 * Component that renders the trail map with gradient lines connecting locations.
 * Draws trails between location nodes with colors based on rank and whether taken.
 */
export class MapComponent extends Component {
  private worldMap: WorldMap;
  private devMode: boolean;

  constructor(width: number, height: number, worldMap: WorldMap, devMode: boolean = false) {
    super(width, height);
    this.worldMap = worldMap;
    this.devMode = devMode;
  }

  override render(ctx: CanvasRenderingContext2D): void {
    if (!this.isVisible()) {
      return;
    }

    ctx.save();

    for (let rank = 0; rank <= this.worldMap.getMaxRank(); rank++) {
      const nodes = this.worldMap.getMapNodes().get(rank) || [];

      for (const node of nodes) {
        const outboundTrails = node.getOutboundTrails();

        for (const trail of outboundTrails) {
          if (trail.isVisible() || this.devMode) {
            this.drawIndividualTrail(
              ctx,
              trail.getOrigin().getPlayerMapX(),
              trail.getOrigin().getPlayerMapY(),
              trail.getDestination().getPlayerMapX(),
              trail.getDestination().getPlayerMapY(),
              node,
              trail
            );
          }
        }
      }
    }

    ctx.restore();

    super.render(ctx);
  }

  private drawIndividualTrail(
    ctx: CanvasRenderingContext2D,
    startX: number,
    startY: number,
    endX: number,
    endY: number,
    node: LocationNode,
    trail: TrailEdge
  ): void {
    ctx.lineWidth = 3;

    const colorMult = node.getRank() / this.worldMap.getMaxRank();
    let startColor: Color;
    let endColor: Color;

    if (trail.isTaken()) {
      startColor = new Color(0, 255, 255);
      endColor = new Color(0, 255, 255);
    } else {
      startColor = new Color(
        Math.floor(100 * colorMult),
        Math.floor(50 * colorMult),
        Math.floor(30 * colorMult)
      );
      endColor = new Color(
        Math.floor(255 * colorMult),
        Math.floor(128 * colorMult),
        Math.floor(64 * colorMult)
      );
    }

    const gradient = ctx.createLinearGradient(
      startX + 5,
      startY + 5,
      endX + 5,
      endY + 5
    );
    gradient.addColorStop(0, startColor.toCSS());
    gradient.addColorStop(1, endColor.toCSS());

    ctx.strokeStyle = gradient;
    ctx.beginPath();
    ctx.moveTo(startX + 5, startY + 5);
    ctx.lineTo(endX + 5, endY + 5);
    ctx.stroke();
  }
}
